documentation/correctness.adoc

```diff
@@ -0,0 +1,133 @@
+The `serializationFormatVersion` property of the root object of the serialization indicates which version to use (or should do so).
+Currently, there's only one version, which is identified as a string such as `"2023.1"`.
+
+Note that JSON Schema isn't powerful enough to express more high-level, nor constraints of a referential nature on the serialization chunk.
+
+Incorrectness is to be reported in terms of e.g. a https://github.com/json-path/JsonPath[JsonPath] to the location(s) of failure(s).
+This is typically dependent on the precise JSON Schema validator framework/library/utility used.
+Recovering from an incorrectness is sometimes possible, e.g.:
+
+. If extra properties are present, they can simply be removed or ignored.
+. If properties are missing, they could just be added, provided they have a compliant default value
+
+=== Hierarchical
+
+The JSON doesn't violate the following constraints.
+
+. IDs of serialized nodes are unique (within the serialization).
+Violation is to be reported in terms of precise locations in the JSON file of all the serialized nodes with equal ID.
+. The `parent` property of a serialized node matches the `id` of the containing node (whenever that containing node is present in the serialization chunk).
+Concretely: if a node `N` is a child of a node `P` then `N.parent` should equal `P`.
+Violation is to be reported in terms of the ID of the contain**ing** serialized node, the ID of the contain**ed** serialized node, and the `parent` ID stated by that.
+. No duplicate containments, i.e. nodes are not the children of multiple other nodes, nor of themselves.
+Violation is to be reported in terms of the ID of the duplicately-contained serialized node, and the IDs of all serialized nodes that claim to contain it, including the links by which this happens, as well the index in the value of those links.
+. No containment cycles, i.e. nodes are not containing themselves.
+. **TODO**  extend for annotations
+
+At this point, links (so: containments and references) don't need to resolve (yet).
+Concretely, this means that IDs that are members of the `children` (for containments) or `targets` (for references) arrays, don't need to resolve to a node in the same JSON serialization.
+They could -- or even: should -- reside in another serialization.
+That pertains specifically to references to language entities of the LIonCore builtins language, which is assumed to be imported/available implicitly.
+If all links do resolve, we say the serialization chunk is _self-contained_.
+
+=== Meta-referential
+
+All the metamodels specified in the serialization chunk should be _made available_ to the client('s deserializer) in the form of a runtime representation that the deserializer can reconstruct from the serializatino.
```

Comment:
I don't think that remarks on deserialization (like here) are relevant.  In LionWeb we define JSON representations (sofar) for language and models and we should only refer to those.  I think everything mentioned in the  document can be done that way. All the validations heere can be done without actually deserializing anything.
The metamodels should therefore be made available in JSON format.

Reply:
That's a good idea: demand that the M2s be available in serialized form.

I think I'd still like to say something about the correctness of deserialized models though, specifically because we also want to say something about reference resolution and constraints, but it wouldn't be bad to separate those aspects more clearly and cleanly.